### DISCUSSION TITLE
Language question : how to express an if/else without computing all branches with select()

### DISCUSSION THREAD

**Discussion by ntimeu** [halide/Halide#6206] [Q&A]
Hello,

At my current work we are trying to port some OpenCL image processing algorithms. On one of them, we are trying to express an if/else, because depending on the value of the pixel (x, y), we either use one algorithm or another one.

Problem, if I look at the select() documentation, it says that " it always evaluates all arguments". Our issue is that both "branches" of the algorithm are quite expensive (N*N convolutions, etc), so we (probably ?) should avoid computing both.

So the question would be : what is the best way to express an if/else in Halide ?

So far, what I'm trying to do is to move the select() close to the smallest number of instructions (moving it to where its execution won't impact that much), but I am wondering if a better way exists.

(I know, premature optimizations are evil, but just looking at the code tell me performances will be horrible with the current implementation).

**Comment by abadams** [comment]
The way to do this in Halide is to exploit the following fact:

If a Func's value is only used when some select condition is true, and that condition has the same value for the lifetime of the Func, then we place the evaluation of the Func inside an if statement and skip it when the condition is false. Here's an example:

```
Func f;
f(x, y) = ...

Func g;
g(x, y) = ...

Func h;
h(x, y) = select(some_condition_that_depends_on_x_y, f(x, y), g(x, y));

// Compute f and g inside all the loop variables the condition depends on
f.compute_at(h, x);
g.compute_at(h, x);
```

If you look at the .stmt file, you should see the realization of f and g guarded by if statements.

**Reply by ntimeu** [reply]
@abadams thanks for the reply. I am going to try it today.

**Reply by ntimeu** [reply]
Hmm ... I'm not sure I have it right. My code look like this :

`
Func condition(x, y) = input(x, y) > CONSTANT;

Func f, g;
Func h;
h(x, y) = select(condition(x, y), f(x, y), g(x, y));
`

The condition depends on pixels value, and I am unsure if I can write an if(condition(x, y)) like that in the code, given the fact that condition(x, y) is an Halide expression, not an actual boolean value (at least not known at compile time).
Also, we are forced to write this algorithm in a generator, to use the autoscheduler. I'm not sure if I can call f.compute_at(h, x) in a generate().


At the same time I am trying to "reverse" our algorithm, to put select() expressions as low as possible, to limit computations on both sides of the select().

**Reply by immortalsalomon** [reply]
Hi, hope you are well. Do you managed to find a solution?

Thank in advance. :)
